src/main/java/remer/sudoku/SudokuGui.java

```diff
@@ -69,10 +70,28 @@ public void changedUpdate(DocumentEvent e) {
         // Add components to the frame
         frame.add(gridPanel, BorderLayout.CENTER);
 
+        // Initialize controller
+        controller = new SudokuController(this, board);
```

Comment:
You're creating the controller, but you're never actually using it in your frame.